### PR TITLE
store/tikv: handle the large transaction commit dead lock

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

In the large transaction protocol, a transaction could update its TTL,
so we can't count on TTL to avoid the deadlock.

For example, in this order:

```
t1 lock k1
t2 lock k2
t1 lock k2
t2 lock k1 ... (deadlock)
```

### What is changed and how it works?

In the prewrite operation, if one transaction meets another transaction lock, we detect a
write-write conflict. One of them should be abort.

Minor changes:
- `maxLockTTL` change to `ManagedLockTTL`, as we can update TTL we don't need such large value.
- fix a typo
- split the resolve lock function into  `ResolveLocks` `resolveLocksForWrite` for read and write

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Write release note for bug-fix or new feature.

Handle deadlock when two large transactions commit and they confict with each other.

### CODE REVIEW

**Comment by tiancaiamao** [comment]
/rebuild

PTAL @youjiali1995 @coocood

File: store/tikv/lock_resolver.go [hunk 1]
```diff
@@ -327,8 +334,15 @@ func (lr *LockResolver) ResolveLocks(bo *Backoffer, callerStartTS uint64, locks
 			// Update the txn expire time.
 			msBeforeLockExpired := lr.store.GetOracle().UntilExpired(l.TxnID, status.ttl)
 			msBeforeTxnExpired.update(msBeforeLockExpired)
-			// In the write conflict scenes, callerStartTS is set to 0 to avoid unnecessary push minCommitTS operation.
-			if callerStartTS > 0 {
+			if forWrite {
+				// Write conflict detected!
+				// If it's a optimistic conflict and current txn is later than the lock owner,
```
Comment (youjiali1995) [inline store/tikv/lock_resolver.go]: current txn is "earlier "than?

File: store/tikv/lock_resolver.go [hunk 2]
```diff
@@ -327,8 +334,15 @@ func (lr *LockResolver) ResolveLocks(bo *Backoffer, callerStartTS uint64, locks
...
+				// abort current transaction.
+				// This could avoids the deadlock scene of two large transaction.
+				if l.LockType == kvrpcpb.Op_Put && l.TxnID > callerStartTS {
```
Comment (youjiali1995) [inline store/tikv/lock_resolver.go]: I think it should abort with any `LockType`.

Comment (tiancaiamao) [reply store/tikv/lock_resolver.go]: 1. The pessimistic lock might rollback, maybe we retry and the pessimistic lock has gone.
2. The pessimistic lock could turn into an optimistic lock in the prewrite phase, so the conflict could be detected anyway

**Comment by youjiali1995** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

File: store/tikv/lock_resolver.go [hunk 3]
```diff
@@ -327,8 +334,15 @@ func (lr *LockResolver) ResolveLocks(bo *Backoffer, callerStartTS uint64, locks
...
 			msBeforeTxnExpired.update(msBeforeLockExpired)
-			// In the write conflict scenes, callerStartTS is set to 0 to avoid unnecessary push minCommitTS operation.
-			if callerStartTS > 0 {
+			if forWrite {
+				// Write conflict detected!
+				// If it's a optimistic conflict and current txn is earlier than the lock owner,
+				// abort current transaction.
+				// This could avoids the deadlock scene of two large transaction.
+				if l.LockType != kvrpcpb.Op_PessimisticLock && l.TxnID > callerStartTS {
+					return msBeforeTxnExpired.value(), nil, kv.ErrWriteConflict.GenWithStackByArgs(callerStartTS, l.TxnID, status.commitTS, l.Key)
```
Comment (cfzjywxk) [inline store/tikv/lock_resolver.go]: Maybe we could add some metrics for this branch, or should we differ the `Resolver` metric `forWrite` version

**Comment by tiancaiamao** [comment]
PTAL @cfzjywxk

**Pull request merged**